src/reference/mod.rs

```diff
@@ -0,0 +1,367 @@
+//! stanzas provide conflicting values for a particular attribute for a particular node or edge.
+//!
+//! # High-level structure
+//!
+//! A graph DSL file consists of one or more **_stanzas_**.  Each stanza starts with a tree-sitter
+//! [**_query pattern_**][query], which identifies a portion of the concrete syntax tree.  The
+//! query pattern should use [**_captures_**][captures] to identify particular syntax nodes in the
+//! matched portion of the tree.  Following the query is a **_block_**, which is a sequence of
+//! statements that construct **_graph nodes_**, link them with **_edges_**, and annotate both with
+//! **_attributes_**.
+//!
+//! [query]: https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries
+//! [captures]: https://tree-sitter.github.io/tree-sitter/using-parsers#capturing-nodes
+//!
+//! Comments start with a semicolon, and extend to the end of the line.
+//!
+//! To execute a graph DSL file against a concrete syntax tree, we execute each stanza in the graph
+//! DSL file _in order_.  For each stanza, we identify each place where the concrete syntax tree
+//! matches the query pattern.  For each of these places, we end up with a different set of nodes
+//! assigned to the query pattern's captures.  We execute the block of statements for each of these
+//! capture assignments, creating any nodes, edges, or attributes mentioned in the block.
+//!
+//! For instance, the following stanza would match all of the identifiers in our example syntax
+//! tree:
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!   ; Some statements that will be executed for each identifier in the
+//!   ; source file.  You can use @id to refer to the matched identifier
+//!   ; node.
+//! }
+//! ```
+//!
+//! # Expressions
+//!
+//! The value of an expression in the graph DSL can be any of the following:
+//!
+//!   - a boolean
+//!   - a string
+//!   - an integer (unsigned, 32 bits)
+//!   - a reference to a syntax node
+//!   - a reference to a graph node
+//!   - an ordered list of values
+//!   - an unordered set of values
+//!
+//! The boolean literals are spelled `#true` and `#false`.
+//!
+//! String constants are enclosed in double quotes.  They can contain backslash escapes `\\`, `\0`,
+//! `\n`, `\r`, and `\t`:
+//!
+//!   - `"a string"`
+//!   - `"a string with\na newline"`
+//!   - `"a string with\\a backslash"`
+//!
+//! Integer constants are encoded in ASCII decimal:
+//!
+//!   - `0`
+//!   - `10`
+//!   - `42`
+//!
+//! # Syntax nodes
+//!
+//! Syntax nodes are identified by tree-sitter query captures (`@name`).  For instance, in our
+//! example stanza, whose query is `(identifier) @id`, `@id` would refer to the `identifier` node
+//! that the stanza matched against.
+//!
+//! # Graph nodes
+//!
+//! Each node in the output graph is identified by a syntax node and a list of **_tag names_**.
+//! Graph nodes are automatically created when they are referenced.  Graph node expressions have
```

Comment:
> Graph nodes are automatically created when they are referenced.

If I'm following correctly, this means that a graph node is automatically created when a _syntax node_ is referenced? I think the word _they_ in this context makes this a bit ambiguous.

Reply:
Each graph node is generated in virtue of being referenced, but also each syntax node is a graph node.